### PR TITLE
Logger Destination Refactor

### PR DESCRIPTION
**Pull request by eracah**
This PR:
* removes `log_data`  from `LoggerDestination` and its subclasses
* adds `log_metrics`, and `log_hyperparameters` to `LoggerDestination` to `FileLogger`, `InMemoryLogger`, `ProgressBarLogger`, `WandBLogger`, and `TensorboardLogger`
* Adds `log_traces` to `FileLogger`
* removes `data`, `data_fit`, `data_epoch`, and `data_batch` from `Logger`
* adds `log_traces`, `log_metrics`, and `log_hyperparameters` to `Logger`
* replaces all `log_data` calls with `log_metrics`, `log_hyperparameters`, or `log_traces` calls
* replaces all `data_fit` calls with `log_hyperparameters`
* replaces all `data_epoch` and `data_batch` calls with `log_metrics`
* removes all `LogLevel` stuff except for in `log_artifacts`
* Along for the ride:
    * moved train metrics before next_batch call, so that the batch is correct for train metrics
    * reordered the hooks in pre-commit, so that pyright is last.
    * configured isort to fix the wandb import bug
* reformats how things are logged. Examples below

###  FileLogger
*  `FileLogger` before: 
```
[FIT][batch=0]: { "blurpool/num_blurpool_layers": 0, "blurpool/num_blurconv_layers": 0, }
[EPOCH][batch=0]: { "epoch": 0, }
[EPOCH][batch=469]: { "metrics/train/Accuracy": 0.8292, }
```
*  `FileLogger` after: 
```
[hyperparameter]: blurpool/num_blurpool_layers: 0 
[hyperparameter]: blurpool/num_blurconv_layers: 0
[metric][batch=0]: trainer/epoch: 0 
[metric][batch=469]: metrics/train/Accuracy: 0.9453 
```

### ProgressBarLogger
* `ProgressBarLogger` log to console before:
```
[FIT]: { "blurpool/num_blurpool_layers": 0, "blurpool/num_blurconv_layers": 0, }
[EPOCH][batch=0/512]: { "epoch": 0, }
[EPOCH][batch=469/512]: { "metrics/train/Accuracy": 0.8292, }
```

* `ProgressBarLogger` log to console after:
```
[hyperparameter]: blurpool/num_blurpool_layers: 0
[hyperparameter]: blurpool/num_blurconv_layers: 0
[batch=0/512]: trainer/epoch: 0
[batch=469/512]: metrics/train/Accuracy: 0.9609
```
* `ProgressBarLogger` locally before:
```
eval           Batch    128:  100%|█████████████████████████| 157/157 [00:01<00:00, 132.27ba/s, metrics/eval/Accuracy=0.8830]                 
eval           Batch    256:  100%|█████████████████████████| 157/157 [00:01<00:00, 127.26ba/s, metrics/eval/Accuracy=0.9390]                 
eval           Batch    384:  100%|█████████████████████████| 157/157 [00:01<00:00, 130.67ba/s, metrics/eval/Accuracy=0.9522]                 
eval           Batch    512:  100%|█████████████████████████| 157/157 [00:00<00:00, 161.95ba/s, metrics/eval/Accuracy=0.9571]                 
train                         100%|█████████████████████████| 512/512 [00:23<00:00, 35.27ba/s, loss/train=0.1819]
```

* `ProgressBarLogger` locally after (no change?):
```
eval           Batch    128:  100%|█████████████████████████| 157/157 [00:01<00:00, 125.96ba/s, metrics/eval/Accuracy=0.9083]         
eval           Batch    256:  100%|█████████████████████████| 157/157 [00:01<00:00, 125.61ba/s, metrics/eval/Accuracy=0.9239]         
eval           Batch    384:  100%|█████████████████████████| 157/157 [00:01<00:00, 120.01ba/s, metrics/eval/Accuracy=0.9494]         
eval           Batch    512:  100%|█████████████████████████| 157/157 [00:01<00:00, 128.42ba/s, metrics/eval/Accuracy=0.9321]         
train                         100%|█████████████████████████| 512/512 [00:23<00:00, 31.37ba/s, loss/train=0.9353] 
```

`ProgressBarLogger` in a notebook before:
Screen Shot 2022-08-17 at 2 16 15 PM

`ProgressBarLogger` in a notebook after (no change?):
Screen Shot 2022-08-17 at 2 17 36 PM

`ProgressbarLogger` from remote logs:
```
  3% 1/29 [00:00<00:20,  1.39ba/s, loss/train=0.4214]
  3% 1/29 [00:01<00:20,  1.39ba/s, loss/train=0.3871]
  7% 2/29 [00:01<00:18,  1.45ba/s, loss/train=0.3871]
  7% 2/29 [00:02<00:18,  1.45ba/s, loss/train=0.3705]
 10% 3/29 [00:02<00:17,  1.49ba/s, loss/train=0.3705]
 10% 3/29 [00:02<00:17,  1.49ba/s, loss/train=0.3852]
 14% 4/29 [00:02<00:16,  1.48ba/s, loss/train=0.3852]
 14% 4/29 [00:03<00:16,  1.48ba/s, loss/train=0.3254]
```

`ProgressBarLogger` in remote interactive session:
Screen Shot 2022-08-23 at 12 07 08 PM

`ProgressBarLogger` `log_to_console` remote:
```
[epoch=2][batch=27/29]: loss/train: 0.1747
[epoch=2][batch=28/29]: trainer/global_step: 86
[epoch=2][batch=28/29]: trainer/batch_idx: 28
[epoch=2][batch=28/29]: trainer/grad_accum: 1
[epoch=2][batch=28/29]: loss/train: 0.1650
[epoch=3][batch=0/10]: epoch: 3
[epoch=3][batch=0/10]: trainer/global_step: 87
[epoch=3][batch=0/10]: metrics/eval/CrossEntropy: 0.1488
[epoch=3][batch=0/10]: metrics/eval/Accuracy: 0.9547
```

* [`WandBLogger` before](https://wandb.ai/mosaic-ml/composer/runs/29iz9rdt)

* [`WandBLogger` after](https://wandb.ai/mosaic-ml/evan-composer/runs/38vjabcx)
* [`WandBLogger` when run on cluster](https://wandb.ai/mosaic-ml/evan-tb-test/runs/1yghi9mb?workspace=user-evan-mosaicml)

* [`TensorboardLogger` before](https://tensorboard.dev/experiment/Wg1sAa1JRCONRAm9bkdIgQ/):
* [`TensorboardLogger` after](https://tensorboard.dev/experiment/KAKaEeC0SVGSqY6B80I4gQ/):

Solves JIRA Issues:
CO-585
CO-916
CO-207

### CODE REVIEW

**Comment by mvpatel2000** [comment]
Why does the before have empty algo charts?

File: tests/metrics/test_current_metrics.py
Comment (mvpatel2000) [inline tests/metrics/test_current_metrics.py:120]: can you put a comment here? what exactly is the indexing doing and why did it change? (Mostly so test is more readable)

**Review by mvpatel2000 — COMMENTED** [review]
This is a massive lift -- huge kudos! No comments on the actual implementation -- it looks the same as the last PR / previous review? so lgtm. Put a few minor comments otherwise.

I'm happy to approve post-tests looking good. Definitely want to make sure we test all the loggers before rolling it out. Maybe add some research folks who use the loggers more often to validate the usage pictures look good? Besides that, this looks good to merge for me.

As a side, it looks like the diff is weird and showing things from another PR I merged into dev. Maybe because this PR hasn't been merged up or something...

**Comment by eracah** [comment]
> Why does the before have empty algo charts?

these were the traces that were getting logged as metrics

File: tests/metrics/test_current_metrics.py
Comment (eracah) [reply tests/metrics/test_current_metrics.py]: fixed here: 
21f75e4 to be more readable

File: tests/loggers/test_in_memory_logger.py
Comment (eracah) [reply tests/loggers/test_in_memory_logger.py:39]: fixed

**Comment by mvpatel2000** [comment]
I'll be OOO, but this lgtm. Not approving because some lint stuff is off, but the logic looks great. Thanks for dealing with some nasty merges

File: composer/loggers/logger.py
Comment (hanlint) [inline composer/loggers/logger.py:130]: Can we keep these around, and throw an error? e.g. `raise NotImplementedError(data_fit is no longer a valid call to the logger API. Please use log_hyperparameters or log_metrics instead)

File: composer/loggers/logger_destination.py
Comment (hanlint) [inline composer/loggers/logger_destination.py:53]: This new API seems a bit limiting? `wandb.log(data)` can take any type of data (time series, images, etc.) yet now through the logger you can _only_ log scalars. 

e.g. https://github.com/mosaicml/composer/blob/dev/composer/callbacks/image_visualizer.py has not been updated in this PR and would break, correct?

File: composer/loggers/logger_destination.py
Comment (hanlint) [inline composer/loggers/logger_destination.py:67]: Not sure if this is actually necessary?

**Review by hanlint — COMMENTED** [review]
Thanks, a significant improvement in our logging API. I am OK with the documentation deferring to a different PR.

But had one significant concern wrt the API (noted in comment above), perhaps we can close offline.

File: composer/loggers/logger.py
Comment (eracah) [reply composer/loggers/logger.py:130]: done

File: composer/loggers/logger_destination.py
Comment (eracah) [reply composer/loggers/logger_destination.py:67]: `log_traces` was added, so that wandb, tb, cometml don't have to log the engine traces which are verbose and only useful if you're using the filelogger or the console logger

File: composer/loggers/logger_destination.py
Comment (eracah) [reply composer/loggers/logger_destination.py:53]: ok fixed by changing `log_metrics` to Any for WandB and InMemoryLogger and calling `log_metrics` for the image visualizer

**Comment by eracah** [comment]
Update:
 * accuracy metrics don't show up in wandb when using yahp entrypoint, but they do show up when using a custom entrypoint
 * the progress bar isn't rendered when running on a cluster 
 * I can't figure out why jenkins is failing
 
 😞 😞 😞 😞 😞

**Comment by eracah** [comment]
> * accuracy metrics don't show up in wandb when using yahp entrypoint, but they do show up when using a custom entrypoint

 fixed: error caused by passing step explicitly to `wandb.log`. If the step value is less than wandb's internal step counting then the data point gets dropped

> * I can't figure out why jenkins is failing

strange isort issue, where I have to manually reorder the imports in `wandb_logger.py`

**Review by mvpatel2000 — APPROVED** [review]
LGTM. Can we open Jira for progressbarlogger logs? 

I suspect this will have some issues we'll discover with internal testing, but no reason to hold this off any longer.

**Review by dskhudia — APPROVED** [review]
As this is already reviewed, accepting to unblock.

File: .pre-commit-config.yaml
Comment (hanlint) [inline .pre-commit-config.yaml:96]: is there a reason for this? is this commit noise?

File: composer/loggers/file_logger.py
Comment (hanlint) [inline composer/loggers/file_logger.py:187]: separate PR: maybe rename the ungainly `format_log_data_value` into something more understandable, e.g. `format_as_string`.

File: composer/loggers/in_memory_logger.py
Comment (hanlint) [inline composer/loggers/in_memory_logger.py:49]: separate PR: maybe rename `InMemoryLogger` -> `DictLogger`

File: composer/trainer/trainer.py
Comment (hanlint) [inline composer/trainer/trainer.py:1498]: not sure if `yapf` likes this, but can we break this up as:
```
self.logger.log_metrics({
    f'metrics/{dataloader_label}{name}': val
    for (name, val) in computed_metrics.items()
}, step=self.state.timestamp.batch.value)
```

File: composer/trainer/trainer.py
Comment (hanlint) [inline composer/trainer/trainer.py:1593]: I think we explicitly wanted `epoch` to not be indented with `trainer` previously (cc.. @abhi-mosaic ? )

File: composer/trainer/trainer.py
Comment (hanlint) [inline composer/trainer/trainer.py:1594]: instead of this, can we do:
```
step=int(self.state.timestamp.batch)
```

Similarly for all the references below

File: composer/trainer/trainer.py
Comment (hanlint) [inline composer/trainer/trainer.py:1843]: does the `logger` have access to state? If so, we can make `step` default to the timestamp to make the code more concise?

File: docs/source/trainer/logging.rst
Comment (hanlint) [inline docs/source/trainer/logging.rst:116]: This docs file needs to have the `Logging Levels` section removed.

**Review by hanlint — CHANGES_REQUESTED** [review]
Made some comments here -- looking good! Using Request Changes to hold this PR until https://github.com/mosaicml/composer/pull/1419 is merged.

File: .pre-commit-config.yaml
Comment (eracah) [reply .pre-commit-config.yaml:96]: just moving pyright to run last to make it easier to surface pyright errors. Otherwise they get hidden by the output of the other precommits. See "along for the ride" section of the pr description

File: composer/loggers/file_logger.py
Comment (eracah) [reply composer/loggers/file_logger.py:187]: captured in CO-971

File: composer/loggers/in_memory_logger.py
Comment (eracah) [reply composer/loggers/in_memory_logger.py:49]: captured in CO-971

File: composer/trainer/trainer.py
Comment (eracah) [reply composer/trainer/trainer.py]: done

File: composer/trainer/trainer.py
Comment (eracah) [reply composer/trainer/trainer.py]: Already defaults to that. Just forgot to remove the optional argument when calling the function

**Review by hanlint — APPROVED** [review]
Approving to unblock now that #1419 is merged

**Comment by eracah** [comment]
> Approving to unblock now that #1419 is merged

🙌

File: composer/trainer/trainer.py
Comment (eracah) [reply composer/trainer/trainer.py]: ok fixed

File: docs/source/trainer/logging.rst
Comment (eracah) [reply docs/source/trainer/logging.rst:116]: done

**Pull request merged**
